src/ecmult_const_impl.h

```diff
@@ -12,220 +12,244 @@
-         * therefore we do not need to check its sign. In particular, it's easy
-         * to see that u_last is never < 0 because u is never < 0. Moreover,
-         * u_last is never = 0 because u is never even after a loop
-         * iteration. The same holds analogously for the initial value of
-         * u_last (in the first loop iteration). */
-        VERIFY_CHECK(u_last > 0);
-        VERIFY_CHECK((u_last & 1) == 1);
-        u += even;
-        u_last -= even * (1 << w);
-
-        /* 4.3, adapted for global sign change */
-        wnaf[word++] = u_last * global_sign;
-
-        u_last = u;
-    } while (word * w < size);
-    wnaf[word] = u * global_sign;
-
-    VERIFY_CHECK(secp256k1_scalar_is_zero(&s));
-    VERIFY_CHECK(word == WNAF_SIZE_BITS(size, w));
-    return skew;
-}
+     * The idea there is to interpret the bits of a scalar as signs (1 = +, 0 = -), and compute a
+     * point multiplication in that fashion. Let v be an n-bit non-negative integer (0 <= v < 2^n),
+     * and v[i] its i'th bit (so v = sum(v[i] * 2^i, i=0..n-1)). Then define:
+     *
+     *   C_l(v, A) = sum((2*v[i] - 1) * 2^i*A, i=0..l-1)
+     *
+     * Then it holds that C_n(v, A) = sum((2*v[i] - 1) * 2^i*A, i=0..l-1)
+     *                              = (2*sum(v[i] * 2^i, i=0..l-1) + 1 - 2^l) * A
+     *                              = (2*v + 1 - 2^l) * A
+     *
+     * Thus, one can compute q*A as C_256((q + 2^l - 1) / 2, A). This is the basis for the
+     * signed-digit multi-comb algorithm for multiplication with a precomputed table.
+     *
+     * It is appealing to try to combine this with the GLV optimization; the idea that a scalar
+     * a can be written as s1 + lambda*s2, where lambda is a curve-specific constant such that
+     * lambda*A is easy to compute, and where s1 and s2 are small. In particular we have the
+     * secp256k1_scalar_split_lambda function which performs such a split with the resulting s1
+     * and s2 in range (-2^128, 2^128) mod n. This does work, but is uninteresting:
+     *
+     *   To compute q*A:
+     *   - Let s1, s2 = split_lambda(q)
+     *   - Let R1 = C_256((s1 + 2^256 - 1) / 2, A)
+     *   - Let R2 = C_256((s2 + 2^256 - 1) / 2, lambda*A)
+     *   - Return R1 + R2
+     *
+     * The issue is that while s1 and s2 are small-range numbers, (s1 + 2^256 - 1) / 2 mod n
+     * and (s2 + 2^256 - 1) / 2 mod n are not, undoing the benefit of the splitting.
+     *
+     * To make it work, we apply the corrections to the input scalar, including the division by
+     * two, *before* calling split_lambda. Let's introduce an offset k, and solve for it. We also
+     * add 2^128 to s1 and s2 after splitting to make sure we end up with non-negative scalars
+     * (a slightly smaller offset would work due to the bounds on the split, but we pick 2^128
+     * for simplicity).
+     *
+     *   To compute q*A:
+     *   - Let s1, s2 = split_lambda((q + k) / 2 mod n)
```

Comment:
It's a bit jarring to see the form of the solution magically appear here as (q + k)/2. It could just be 't' (or 'f(q)') and the derivation below works out fine, but without the leap.